Search: Fix body scroll for overlay open/close

kangzj opened the pull request:
#### Changes proposed in this Pull Request:
The PR fixed an issue when user closes overlay the page scrolls to the top introduced by #20678. And it also includes some refactoring, which extracted some more dom integration functionalities.

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
* Go to a site with Jetpack Search subscription on mobile device
* Perform a search, make sure the search input is visible 
* Ensure the body in the background is not scrollable
* Ensure body is not scrolled to the top

* Close the overlay
* Ensure the page is not scrolled to the top

Known issue: if the overlay is triggered by submit, the search input on the page would lose focus after closing the overlay.

kangzj commented on projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx:
```diff
@@ -183,8 +183,12 @@ export default class DomEventHandler extends Component {
+		// NOTE: the propertyName need to be aligned with the animation.
+		// We don't scroll the overlay on overlay close.
+		if (
+			event.propertyName === 'opacity' &&
```
it should be `===` but not `!==` here, which was a typo

bluefuton replied:
Sorry I missed this in the review! 👀

kangzj replied:
@bluefuton nah mate, you are all good. This is my silly typo which shouldn't happen 😄 .

bluefuton commented on projects/plugins/jetpack/modules/search/instant-search/components/overlay.scss:
```diff
@@ -21,7 +21,8 @@ $overlay-horizontal-padding-lg: 3em;
 	overflow-y: auto;
-	
+
+	// If the animation is to be changed, please do change the `propertyName` in function `scrollOverlayToTop` too.
```
Future developer thanks you!

bluefuton reviewed (dismissed):
On my test site, I'm noticing that when I use a search input further down the page, the underlying page scrolls to top once the modal is closed.

kangzj commented:
> On my test site, I'm noticing that when I use a search input further down the page, the underlying page scrolls to top once the modal is closed.

Hi @bluefuton, sorry this is now fixed. It happened because the conditions aren't quite right sometimes.

And also I noticed another problem - if the trigger is set to `submission`, when closing the overlay, the search input wouldn't be focused as well. So fixed this as well, which makes the PR a bit bigger than I firstly expected.

bluefuton commented:
> And also I noticed another problem - if the trigger is set to submission, when closing the overlay, the search input wouldn't be focused as well. So fixed this as well, which makes the PR a bit bigger than I firstly expected.

As discussed, let's break this out to another PR if possible to make life easier for the reviewer :)

kangzj committed:
[not verified] fixed condition

kangzj committed:
[not verified] changelog

kangzj commented:
> As discussed, let's break this out to another PR if possible to make life easier for the reviewer :)

Hi @bluefuton thanks for the suggestion. I only left the code to address the original issue now :-)

kangzj commented on projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx:
```diff
@@ -183,11 +183,13 @@ export default class DomEventHandler extends Component {
-		if ( event?.propertyName !== 'opacity' ) {
+		// NOTE: the propertyName need to be aligned with the animation: instant-search/components/overlay.scss.
+		// We don't scroll the overlay on overlay close.
+		if ( event.propertyName !== 'opacity' || ! this.props.isVisible ) {
```
And so sorry to make you feel bad about this line - it was right in the previous PR, because we want to return early if it's not `opacity`... :neckbeard:

jsnmoon commented on projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx:
```diff
@@ -183,11 +183,13 @@ export default class DomEventHandler extends Component {
-		// @see https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollTo
+
+		// @see https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollTo
 		window?.scrollTo( 0, 0 );
```
We likely want to scroll the overlay element, not the window. See [here](https://github.com/Automattic/jetpack/blame/e76b6ebd438259b44ee2ead459aa4fb3b4410192/projects/plugins/jetpack/modules/search/instant-search/components/search-app.jsx#L170) for how we used to perform overlay scrolls in the past. We won't need to check for overlay visibility using this approach.

jsnmoon replied:
I looked into the original issue more in-depth and discovered that scrolling the overlay element won't fix our issue here. 😢

kangzj replied:
Thanks for looking into this @jsnmoon

jsnmoon reviewed (changes requested):
I'm not sure why we're scrolling the page behind the overlay back and forth when toggling the overlay. This is what I see on the desktop:

1. Using the 2020 theme with a search input in the footer sidebar, open `/`.
2. Scroll all the way down to the footer and type a query into the input from (1).
3. Notice that the page scrolls to the top while the modal fades into view.
4. Close the overlay.
5. Notice that the page scroll returns to the position from (2).

Instead of scrolling the page, I believe we should be scrolling the overlay element. More details inline.

kangzj commented:
> I'm not sure why we're scrolling the page behind the overlay back and forth when toggling the overlay. This is what I see on the desktop:

Hi @jsnmoon Good catch, didn't notice this. Just had a look, this is a side-effect by make the body fixed to prevent it from scrolling on touch events. I guess I'll need to dive into this a bit more to find out a better way. The PR introduce this is #20677.

> Instead of scrolling the page, I believe we should be scrolling the overlay element. More details inline.

`element.scrollTo` just doesn't work here - tbh I have no idea why it worked for the Firefox issue before, mystery. I did some experiments here, you could notice `element.scrollTo` would scroll the content inside an element which is not the case for our issue here.

https://d.pr/i/xFFiK0
code:
```html
<html>
	<body>
		<div style="height: 400px; width: 400px; overflow: scroll;" id="this-is-a-div">
			<div
				style="
					height: 800px;
					width: 100%;
					background: linear-gradient(
						0deg,
						rgba( 0, 212, 255, 1 ) 0%,
						rgba( 2, 0, 36, 1 ) 100%
					);
				"
			></div>
		</div>
	</body>
</html>

```

kangzj committed:
[not verified] move dom integartion out and fix scroll issue

kangzj commented:
Hi @jsnmoon I've fixed the scrolling up and down issue.

We are trying to solve two issues here.
* On mobile devices i.e. when scrolling the overlay using fingers, the body scrolls too #20271 

It works fine on Desktop browsers. But on mobile device, the browser uses touch events to scroll, and the scroll event would bubble up to body, which leads to problem. #20677 fixed the issue by make the body position `fixed`, but also introduced a side effect, which you pointed out, the page would scroll to top. I fixed this in be5b781 by remembering the position on open and setting the scroll position on overlay close. The code does look a bit hacky, I'm all ears to suggestions.

The scrolling control functions are moved to dom integration component, which I think would be clearer.

* On iPad/iPhone, sometimes when overlay pops up, the search input is not visible #19882 

How and why the issue is resolved is explained in the previous comment.

If you still have any questions, just let me know. Thanks 👍

kangzj committed:
[not verified] changelog

bluefuton commented on projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx:
```diff
@@ -182,15 +183,52 @@ export default class DomEventHandler extends Component {
+		window.scrollTo( 0, this.top );
+	}
+
+	getWpAdminBarComputedHeight() {
```
Can we add some comments explaining why this wp-admin bar needs to be taken into account?

Would we face the same issue if the customer site had added their own fixed bar at the top of the page?

kangzj replied:
>Can we add some comments explaining why this wp-admin bar needs to be taken into account?

This is a good question. The problem is wp admin bar offset html to be sticky, which is very rare nowadays. That's the reason why we consider this separately. Otherwise the page would not keep at the same position because of the offset. But I'm sure there's better way to get the offset.

>Would we face the same issue if the customer site had added their own fixed bar at the top of the page?

I did consider this problem. And looked at the P2 bar, which is fine, not like the wp admin bar which does offset to html...

kangzj replied:
changed to use `document.documentElement.scrollHeight - document.body.scrollHeight`

bluefuton commented on projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx:
```diff
@@ -182,15 +183,52 @@ export default class DomEventHandler extends Component {
+	}
+
+	getWpAdminBarComputedHeight() {
+		const $wpAdminBar = document.getElementById( WP_ADMIN_BAR_ID );
```
Should we have the `$` at the start of this var name?

kangzj replied:
Old habit of add $ to a variable holding a node, will change this.

kangzj replied:
The code is not necessary anymore.

bluefuton commented on projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx:
```diff
@@ -182,15 +183,52 @@ export default class DomEventHandler extends Component {
+
+		if ( this.props.isVisible ) {
+			this.preventBodyScroll();
+			// This is to fix search input on the overlay is not visible on mobile devices.
```
```suggestion
			// This ensures the search input is visible on mobile devices.
```

kangzj committed:
[not verified] remove unused function

kangzj committed:
[not verified] Update projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx

Co-authored-by: Chris R <chris.rosser@automattic.com>

kangzj committed:
[not verified] changed fixBodyScroll to more robust lifecycle method

kangzj committed:
[not verified] remove unused code

kangzj committed:
[not verified] added comments

kangzj committed:
[not verified] comment

kangzj committed:
height set to 0 if couldnot read it

jsnmoon commented on projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx:
```diff
@@ -182,15 +177,54 @@ export default class DomEventHandler extends Component {
+		document.body.style.left = '';
+		document.body.style.right = '';
+		document.body.style.position = '';
+		// Restore body position.
+		this.top > 0 && window.scrollTo( 0, this.top );
+		this.top = 0;
```
I'm currently seeing the following behavior when I try this in both desktop and mobile viewports:

1. Using the 2020 theme with a search input in the footer sidebar, open `/`.
2. Scroll all the way down to the footer and type a query into the input from (1).
3. The overlay opens as expected. The body scroll is affixed in the background, which is also expected.
4. Close the overlay.
5. The body abruptly scrolls to the very top and then smoothly scrolls back down to the position from (2).

To fix the abrupt scrolling in (5), we want to temporarily disable `smooth-scroll` on the `html` element like so:

```suggestion
		// Restore body position.
		const previousValue = document.querySelector( 'html' ).style[ 'scroll-behavior' ];
		document.querySelector( 'html' ).style[ 'scroll-behavior' ] = 'revert';
		this.top > 0 && window.scrollTo( 0, this.top );
		document.querySelector( 'html' ).style[ 'scroll-behavior' ] = previousValue ?? '';
		this.top = 0;
```

kangzj replied:
Hi @jsnmoon  wondering if we should use ` document.documentElement.style.scrollBehavior` instead of ` document.querySelector( 'html' ).style[ 'scroll-behavior' ]` because style is an instance of `CSSStyleDeclaration` and `documentElement` looks a bit nicer 😄

jsnmoon replied:
> Hi @jsnmoon wondering if we should use ` document.documentElement.style.scrollBehavior` instead of ` document.querySelector( 'html' ).style[ 'scroll-behavior' ]` because style is an instance of `CSSStyleDeclaration` and `documentElement` looks a bit nicer 😄

If that fixes the issue, then sure! I made my initial suggestion because Chrome's default stylesheet defines the `scroll-behavior` value on the root `html` element instead of the document.

kangzj replied:
![image](https://user-images.githubusercontent.com/1425433/131419599-5dcc641c-d4e7-45ba-a055-c44c1f029d3a.png)
The fix doesn't work for twentytwenty as it sets an `important` body `position`.... 😭

kangzj replied:
This is now fixed.

bluefuton replied:
`!important !important !no-really-this-is-more-important`

jsnmoon reviewed (commented):
Left two actionable suggestions for the PR inline. Thanks for working on this fix, @kangzj!

kangzj committed:
changed top to a state

kangzj commented on projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx:
```diff
@@ -182,15 +177,54 @@ export default class DomEventHandler extends Component {
+	 * 3) And we remember the body postition in `this.top`
+	 */
+	preventBodyScroll() {
+		this.top = parseInt( window.scrollY ) || 0;
```
Thanks 😄  Changed in 0583405e2f.
I did consider to put it as a state, but considering it would introduce more complexity then used a variable. But it surely would look more consistent using a state.

kangzj committed:
fixed case

kangzj committed:
fixed for twentytwenty theme

kangzj committed:
make it more robust

kangzj committed:
comment

kangzj commented:
Hey @jsnmoon would mind having another look please?

jsnmoon commented on projects/plugins/jetpack/modules/search/instant-search/components/dom-event-handler.jsx:
```diff
@@ -22,6 +17,10 @@ export default class DomEventHandler extends Component {
 			// We toggle isComposing on compositionstart and compositionend events.
 			// (CJK = Chinese, Japanese, Korean; see https://en.wikipedia.org/wiki/CJK_characters)
 			isComposing: false,
+			// `bodyScrollTop` remebers the body scroll position.
```
```suggestion
			// `bodyScrollTop` remembers the body scroll position.
```

:)

jsnmoon reviewed (dismissed):
Looks good!

jsnmoon committed:
Fix typo

jeherve reviewed (approved):
This tests well for me. 🚢

kangzj commented:
🚢 shipped in r231028-wpcom
